### ISSUE TITLE
Add logger class

### ISSUE THREAD

**Issue by hsorby** [cellml/libcellml#115]
Add logger class to report errors with parsing a model, would also be used for logging messages when validating models.  This logger would only log messages to memory.

In the current context it is expected that only strings would be logged but then looking to extend it at a later time to maybe include object references etc.

**Comment by agarny** [comment]
Not maybe, it will have to or it will be of no use to a CellML environment like OpenCOR.

**Comment by hsorby** [comment]
I have put together sample of how this could look here:

https://github.com/dladd/libcellml/pull/3

Thoughts?

**Comment by nickerso** [comment]
Looks like a good start to me.

**Comment by dladd** [comment]
Another issue that the logger could potentially handle is what to do with any non-CellML elements and comments we find during parsing. We can similarly log instances of these in the entities in which they occur for use by an interested user or environment.

**Comment by dladd** [comment]
To update on this, in the current work on the logger in pull #114 , there is essentially a mini-object container for every error type we may need. For example, to log errors pertaining to specification of a base unit:

In [parser.cpp](https://github.com/cellml/libcellml/pull/114/files#diff-bd8f167953cc43fb7cba7d420fbd1d48R173), we check the validity of a base_unit attribute and create a `UnitsBaseUnitAttributeError` if we spot an error:

``` c++
...
        } else if (attribute->isType("base_unit")) {
            if (attribute->getValue() == "yes") {
                units->setBaseUnit(true);
            } else if (attribute->getValue() == "no") {
                units->setBaseUnit(false);
            } else {
                UnitsBaseUnitAttributeErrorPtr err = std::make_shared<UnitsBaseUnitAttributeError>();
                err->setUnits(units);
                err->setValue(attribute->getValue());
                addError(err);
            }
...
```

In [loggererrors.cpp](https://github.com/cellml/libcellml/pull/114/files#diff-b0494300bf656936ddfe1a6226349efeR1), we have methods to serialise out a useful message:

``` c++
 std::string UnitsBaseUnitAttributeError::doSerialisation() const
{
    std::string s = "Unrecognised base_unit attribute in units '" + getUnits()->getName() + "': should be either 'yes' or 'no', and not '" + getValue() + "'";
    return s;
}
```

So that error has access to an object appropriate to the error. A user wanting to get this error will need to `getError()` (see [logger.cpp](https://github.com/cellml/libcellml/pull/114/files#diff-0e66cd02ff1f1a28015f74afd818e270R71)) from the errors in the parser and dynamic cast it to the appropriate error object. From there, they can do `getUnits()` if they want to get the actual units object the error is relevant to.

**Comment by agarny** [comment]
This looks like a very good start to me.

**Comment by hsorby** [comment]
This has been implemented in pull request #114 and should now be closed.
